Port CPU torch.orgqr to ATen

IvanYashchuk opened the pull request:
Now we can remove `_th_orgqr`!

Compared to the original TH-based `orgqr`, complex (#33152) and batched inputs are now supported.
CUDA support will be added in a follow-up PR.

Closes https://github.com/pytorch/pytorch/issues/24747

Ref. #49421, #42666

IvanYashchuk committed:
wip porting orgqr to ATen

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into port-orgqr

IvanYashchuk committed:
Don't resize infos, fixed a few asserts

IvanYashchuk committed:
Added tests

IvanYashchuk committed:
Allow incorrectly sized out tensors; added more tau input checks

IvanYashchuk committed:
More tests for errors and warnings

IvanYashchuk committed:
AT_ERROR -> TORCH_CHECK

IvanYashchuk committed:
Use declared/defined dispatch instead of registering helper function in
native_functions.yml

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -982,40 +982,56 @@ static void apply_geqrf(Tensor& self, Tensor& tau, int64_t m, int64_t n,
-    if (info != 0) {
-      return;
-    }
+    lapackOrgqr<scalar_t>(m, n_columns, k, self_working_ptr, lda, tau_working_ptr, work.data_ptr<scalar_t>(), lwork, info_working_ptr);
```
Should this still short-circuit for now since we don't return infos?

IvanYashchuk replied:
I added it back.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1113,6 +1130,115 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+  TORCH_INTERNAL_ASSERT(result.device() == input.device());
+
+  TORCH_INTERNAL_ASSERT(infos.scalar_type() == at::kInt);
+  TORCH_INTERNAL_ASSERT(infos.numel() == std::max<int64_t>(1, batchCount(input)));
```
Does this also need a assertion for the device of infos?

IvanYashchuk replied:
Yep, I added it now.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1113,6 +1130,115 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+    at::native::resize_as_(result, input.transpose(-2, -1), MemoryFormat::Contiguous);
+    result.transpose_(-2, -1);
+  } else {
+    // Resize messes up the strides and we expect strictly column major order, so let's not use at::native::resize_output
```
We should use resize_output for now instead of hard erroring in this case.

IvanYashchuk replied:
Yes, this comment shouldn't be here now. I removed it and moved the assert out of the "else" branch.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1113,6 +1130,115 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+
+  TORCH_CHECK(tau.scalar_type() == input.scalar_type(),
+    "orgqr: tau dtype ", tau.scalar_type(), " does not match input dtype ", input.scalar_type());
+  TORCH_CHECK(input.device() == input.device(),
```
`input.device() == tau.device()`

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1113,6 +1130,115 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+  auto infos = at::empty({std::max<int64_t>(1, batchCount(input))}, input.options().dtype(kInt).device(kCPU));
+
+  // if result is not empty and not in batched column major format we have to allocate a temporary tensor
+  if (result.numel() != 0 && !result.transpose(-2, -1).is_contiguous()) {
```
This is really nice.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1113,6 +1130,115 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+
+Tensor orgqr(const Tensor& input, const Tensor& tau) {
+  Tensor result = at::empty({0}, input.options());
+  result = at::orgqr_outf(input, tau, result);
```
`orgqr_outf` must be a typo here, right?

IvanYashchuk replied:
No, that's the actual function now.
`at::orgqr_outf(input, tau, result)` is equivalent to `at::orgqr_out(result, input, tau)`.

mruberry commented on test/test_linalg.py:
```diff
@@ -4049,21 +4049,98 @@ def test_renorm_ps(self, device):
     @skipCPUIfNoLapack
-    def test_orgqr_errors(self, device):
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    def test_orgqr(self, device, dtype):
```
Pretty cool that this was never tested previously.

mruberry commented on test/test_linalg.py:
```diff
@@ -4049,21 +4049,98 @@ def test_renorm_ps(self, device):
+        shapes = [(0, 0), (5, 0),  # Empty matrix
+                  (5, 5), (5, 3),  # Single matrix
+                  (2, 5, 5), (2, 5, 3),  # 3-dim Tensors
+                  (2, 1, 5, 5), (2, 1, 5, 3)]  # 4-dim Tensors
```
Would you add a zero batch case, too?

IvanYashchuk replied:
Added it now and discovered that `linalg.qr` doesn't work for such inputs https://github.com/pytorch/pytorch/issues/50576.

mruberry commented on test/test_linalg.py:
```diff
@@ -4049,21 +4049,98 @@ def test_renorm_ps(self, device):
+            actual = torch.orgqr(reflectors, tau)
+            self.assertEqual(expected, actual)
+
+            out = torch.empty_like(A)
```
Let's add an OpInfo for this function either in the CUDA port PR or after it. Then we won't need this out test, I think.

mruberry reviewed (commented):
Hey @IvanYashchuk!

This port looks pretty good. @ngimel and I made a few comments.

However, we were looking at this function and wondering if it should just be removed. We could find no uses of it within Facebook. Do you think it's interesting vs. torch.linalg.qr? If we want to support this functionality, maybe we'd offer it as a mode on torch.linalg.qr instead of this cryptically named function.

Looking forward to hearing your thoughts.

EDIT: We preemptively edited the tracking issue https://github.com/pytorch/pytorch/issues/49421 to reflect deprecating and removing, not porting, functions like orgqr. If we decide we don't want to keep it then we can change the issue back.

mruberry commented:
Some more context on deprecating this function, it is used in one function:

```
def basis(A):
    """Return orthogonal basis of A columns.
    """
    if A.is_cuda:
        # torch.orgqr is not available in CUDA
        Q, _ = torch.qr(A, some=True)
    else:
        Q = torch.orgqr(*torch.geqrf(A))
    return Q
```

that appears to be copied around some repos on Github.

IvanYashchuk commented:
> However, we were looking at this function and wondering if it should just be removed. We could find no uses of it within Facebook. Do you think it's interesting vs. torch.linalg.qr? If we want to support this functionality, maybe we'd offer it as a mode on torch.linalg.qr instead of this cryptically named function.

There is a recent request for this function to support CUDA and differentiation https://github.com/pytorch/pytorch/issues/50104. I definitely agree that we shouldn't use lapack's name. I was thinking to introduce a more descriptive name together with the PR for the backward rule, something like `householder_orthogonal` and `householder_orthogonal_mul` for `torch.ormqr`. Maybe this function was not used much exactly because it lacks CUDA and autograd support.

mruberry commented:
> There is a recent request for this function to support CUDA and differentiation #50104. I definitely agree that we shouldn't use lapack's name. I was thinking to introduce a more descriptive name together with the PR for the backward rule, something like `householder_orthogonal` and `householder_orthogonal_mul` for `torch.ormqr`. Maybe this function was not used much exactly because it lacks CUDA and autograd support.

Aha! Thank you for reminding me. We can work on a name improvement like you've suggested in a future PR (using an alias).

mruberry commented:
Linker failures on mobile seem real:

```
Undefined symbols for architecture x86_64:
  "void at::native::apply_orgqr<float>(at::Tensor&, at::Tensor const&, at::Tensor&, long long)", referenced from:
      at::native::(anonymous namespace)::orgqr_kernel_impl(at::Tensor&, at::Tensor const&, at::Tensor&, long long) in libtorch_cpu.a(BatchLinearAlgebraKernel.cpp.o)
  "void at::native::apply_orgqr<double>(at::Tensor&, at::Tensor const&, at::Tensor&, long long)", referenced from:
      at::native::(anonymous namespace)::orgqr_kernel_impl(at::Tensor&, at::Tensor const&, at::Tensor&, long long) in libtorch_cpu.a(BatchLinearAlgebraKernel.cpp.o)
  "void at::native::apply_orgqr<c10::complex<double> >(at::Tensor&, at::Tensor const&, at::Tensor&, long long)", referenced from:
      at::native::(anonymous namespace)::orgqr_kernel_impl(at::Tensor&, at::Tensor const&, at::Tensor&, long long) in libtorch_cpu.a(BatchLinearAlgebraKernel.cpp.o)
  "void at::native::apply_orgqr<c10::complex<float> >(at::Tensor&, at::Tensor const&, at::Tensor&, long long)", referenced from:
      at::native::(anonymous namespace)::orgqr_kernel_impl(at::Tensor&, at::Tensor const&, at::Tensor&, long long) in libtorch_cpu.a(BatchLinearAlgebraKernel.cpp.o)
ld: symbol(s) not found for architecture x86_64
```

and 

```
Jan 13 21:44:47 BatchLinearAlgebraKernel.cpp:(.text+0x14c): undefined reference to `void at::native::apply_orgqr<c10::complex<double> >(at::Tensor&, at::Tensor const&, at::Tensor&, long)'
Jan 13 21:44:47 BatchLinearAlgebraKernel.cpp:(.text+0x161): undefined reference to `void at::native::apply_orgqr<float>(at::Tensor&, at::Tensor const&, at::Tensor&, long)'
Jan 13 21:44:47 BatchLinearAlgebraKernel.cpp:(.text+0x1e4): undefined reference to `void at::native::apply_orgqr<c10::complex<float> >(at::Tensor&, at::Tensor const&, at::Tensor&, long)'
Jan 13 21:44:47 BatchLinearAlgebraKernel.cpp:(.text+0x1f4): undefined reference to `void at::native::apply_orgqr<double>(at::Tensor&, at::Tensor const&, at::Tensor&, long)'
```

Let me know if I should ping someone on the mobile team.

IvanYashchuk commented:
It's the first time I use DEFINE_DISPATCH/DECLARE_DISPATCH and I may have done something wrong, but all other builds work. Hints would be appreciated since I'm totally not familiar with mobile builds.

IvanYashchuk committed:
Add back early return if info != 0

mruberry commented:
> It's the first time I use DEFINE_DISPATCH/DECLARE_DISPATCH and I may have done something wrong, but all other builds work. Hints would be appreciated since I'm totally not familiar with mobile builds.

@dreiss @ezyang -- do you have an idea of why this might be hitting linker errors?

IvanYashchuk committed:
Added assertion for the device of infos

IvanYashchuk committed:
Moved result.sizes()==input.sizes() assert out of else

IvanYashchuk committed:
Fixed input tau device assert

IvanYashchuk committed:
Added zero batch dimension test cases

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into port-orgqr

IvanYashchuk committed:
Trying to fix mobile builds; moved implementation of apply_orgqr to the
header

IvanYashchuk commented:
CI failures are not related to this PR.
Mobile builds now pass. The problem was in not having the implementation of the templated function `apply_orgqr` in the header.
@mruberry this is ready for another round. I addressed the comments.

mruberry reviewed (approved):
Thanks @IvanYashchuk!

mruberry reviewed (approved):
Thanks @IvanYashchuk!

mruberry commented:
This is hitting a series of errors internally. I suggest we rebase and resubmit it using ci-all, especially since this base looks like it has several failing builds. The failures are the CPU variants of test_orgqr:

caffe2/test:linalg - test_orgqr_cpu_float32 (test_linalg.TestLinalgCPU)
caffe2/test:linalg - test_orgqr_cpu_float64 (test_linalg.TestLinalgCPU)
caffe2/test:linalg - test_orgqr_cpu_complex64 (test_linalg.TestLinalgCPU)
caffe2/test:linalg - test_orgqr_cpu_complex128 (test_linalg.TestLinalgCPU)

Example error output:

```
RuntimeError: orgqr: Argument 8 has illegal value
  File "/usr/local/fbcode/platform007/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/local/fbcode/platform007/lib/python3.7/unittest/case.py", line 628, in run
    testMethod()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/opt/gen/caffe2/test/linalg#binary,link-tree/torch/testing/_internal/common_device_type.py", line 286, in instantiated_test
    raise rte
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/opt/gen/caffe2/test/linalg#binary,link-tree/torch/testing/_internal/common_device_type.py", line 281, in instantiated_test
    result = test_fn(self, *args)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/opt/gen/caffe2/test/linalg#binary,link-tree/torch/testing/_internal/common_device_type.py", line 678, in only_fn
    return fn(slf, device, *args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/opt/gen/caffe2/test/linalg#binary,link-tree/torch/testing/_internal/common_device_type.py", line 572, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/opt/gen/caffe2/test/linalg#binary,link-tree/test_linalg.py", line 4100, in test_orgqr
    run_test(shape)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/opt/gen/caffe2/test/linalg#binary,link-tree/test_linalg.py", line 4080, in run_test
    actual = torch.orgqr(reflectors, tau)
```

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into port-orgqr

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.h:
```diff
@@ -15,10 +17,73 @@ namespace at { namespace native {
+  // Since we deal with batches of matrices with the same dimensions, doing this outside
+  // the loop saves (batch_size - 1) workspace queries which would provide the same result
+  // and (batch_size - 1) calls to allocate and deallocate workspace using at::empty()
+  int lwork = -1;
+  scalar_t wkopt;
+  lapackOrgqr<scalar_t>(m, n_columns, k, self_data, lda, tau_data, &wkopt, lwork, &infos_data[0]);
+  lwork = static_cast<int>(real_impl<scalar_t, value_t>(wkopt));
```
I don't like this error `Argument 8 has illegal value`. The 8-th argument is `lwork`, it is an integer for which we should obtain the value from LAPACK.
> The size of the work array (lwork≥n).

MKL doc link: https://software.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem-routines/lapack-least-squares-and-eigenvalue-problem-computational-routines/orthogonal-factorizations-lapack-computational-routines/orgqr.html

IvanYashchuk replied:
@mruberry ci-all is all green (except [this](https://app.circleci.com/pipelines/github/pytorch/pytorch/261921/workflows/6ee88bb2-78c5-4545-bd9e-fae39b278758/jobs/10265138) which seems unrelated).
Could you somehow check what BLAS/LAPACK implementation PyTorch in that failing build is linked to? Is it MKL?`torch.__config__.show()`

mruberry replied:
The versioned used internally seems to be much older than the version in CI.

Could it be the change in how the function is called, using m instead of lda? I'll try to debug internally, too, to get a better sense for what's going on. @ngimel points out that torch.linalg.qr must be relying on this same function, so it's surprising we haven't seen this issue previously.

ngimel commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1113,6 +1073,114 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+    at::native::resize_as_(result, input.transpose(-2, -1), MemoryFormat::Contiguous);
+    result.transpose_(-2, -1);
+  }
+
```
adding early return here 
```
  //early return for empty matrices
  if (result.numel() == 0) {
    infos.fill_(0);
    return result;
  }
  ```
  fixes internal error. I now wonder how OSS tests pass, because for empty matrix lwork is returned as 0, and that's an illegal value (it should be at least 1)

IvanYashchuk replied:
Reference LAPACK for empty matrices returns lwork as 1
https://github.com/Reference-LAPACK/lapack/blob/master/SRC/dorgqr.f#L193-L196
Maybe the older version of MKL didn't do that and now it does the same as in reference implementation, that's why OSS could be passing.
We haven't seen issues with torch.linalg.qr because the early return is used there.

IvanYashchuk committed:
Added early return for empty matrices

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into ci-all/ivanyashchuk/port-orgqr

IvanYashchuk committed:
Removed use_c10_dispatcher: full; it is now default

IvanYashchuk commented:
@mruberry, @ngimel I added an early return for empty matrices. It should fix the internal tests.
